// @ijl/cli/src/webpack/base/index.js
const webpack = require("webpack");
const path = require("path");
const { CleanWebpackPlugin } = require("clean-webpack-plugin");
const webpackCopy = require("copy-webpack-plugin");
const { merge } = require("webpack-merge");

const getProjectConfig = require("../../server/utils/get-project-config");

const { webpackConfig = {} } = getProjectConfig();

const outputDirectory = "dist";

const cwd = process.cwd();

module.exports = merge(
  {
    output: {
      filename: "[name].js",
      path: path.resolve(cwd, outputDirectory),
      libraryTarget: "umd",
      globalObject: `(typeof self !== 'undefined' ? self : this)`
    },
    plugins: [
      new CleanWebpackPlugin(),
      new webpackCopy({
        patterns: [{ from: "./static/", to: "static", noErrorOnMissing: true }],
      })
    ],
    resolve: {
      modules: [cwd, "node_modules"],
      extensions: [".webpack.js", ".web.js", ".ts", ".tsx", ".js", ".css"]
    },

    module: {
      rules: [
        {
          test: /\.tsx?$/,
          loader: "ts-loader"
        },
        {
          test: /\.(jpe?g|gif|png|svg|woff|ttf|eot|wav|mp3)$/,
          loader: "file-loader"
        }
      ]
    },
    // externals: {
    //   react: "react",
    //   "react-dom": "react-dom",
    //   redux: "redux",
    //   "react-redux": "react-redux",
    //   "styled-components": "styled-components"
    // }
  },
  webpackConfig
);
